### ISSUE TITLE
Accuracy drop : DeiT-tiny & LLaMA behavior differences between FP / Decomposition / FQ / TOSA

### ISSUE THREAD

**Issue by hth000** [pytorch/executorch#16426]
### 🐛 Describe the bug

Hi Executorch ARM backend team,

I’m validating the lowering pipeline and collected the following comparison across several models. Quantize with 8-bit only.
Decompose with aten: Decompose by arm backend **`quantizer.transform_for_annotation`**
| Model           | **FP vs Decompose with aten**                                | **FP vs Decompose with quant aten (FQ)**                     | **FP vs TOSA**                                               | **Compare FQ vs TOSA**                                       |
| --------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **Mobilenetv2** | max error: 0.0  <br> mean error: 0.0 <br> cosine: 1.0        | max error: 8.0 <br> mean error: 1.187 <br> cosine: 0.9997984690887474 | max error: 13.0 <br> mean error: 1.731 <br> cosine: 0.9995700057524048 | max error: 8.0 <br> mean error: 1.226 <br> cosine: 0.9997850644158716 |
| **Resnet18**    | max error: 0.0  <br> mean error: 0.0 <br> cosine: 0.9999999999999999 | max error: 6.0 <br> mean error: 1.525 <br> cosine: 0.9990599672781051 | max error: 9.0 <br> mean error: 1.815 <br> cosine: 0.9985450487548194 | max error: 7.0 <br> mean error: 1.49 <br> cosine: 0.9990350537962956 |
| **Deit-tiny**   | max error: 0.0  <br> mean error: 0.0 <br> cosine: 1.0        | max error: 96.0 <br> mean error: 21.23 <br> cosine: 0.8994378800187951 | max error: 83.0 <br> mean error: 18.682 <br> cosine: 0.9264412351657058 | **max error: 96.0 <br> mean error: 21.434 <br> cosine: 0.898735675973962** |
| **LLaMA**       | **max error: 89.0  <br> mean error: 20.6328 <br> cosine: 0.7638448052472464** | max error: 82.0 <br> mean error: 19.37109375 <br> cosine: 0.7869657464752766 | max error: 82.0 <br> mean error: 19.37109375 <br> cosine: 0.7869657464752766 | max error: 0.0 <br> mean error: 0.0 <br> cosine: 1.0         |

With the experiments, the variables are as following.

- Executorch commit: **913436a4**
- TOSA results are inferencing by **tosa_reference_model**
- For all models, test input is *torch.rand(input_shape)* (batch size:1)
- For verification only, calibration data == testing input. 
- The models are using **executorch/example/models**
- LLaMA weights:

```python
torch.manual_seed(0)
for p in self.model_.parameters():
# p.data.fill_(0)
torch.nn.init.normal_(p, mean=0.0, std=0.02)
for b in self.model_.buffers():
b.data.fill_(0)
```

Based on the conditions above, I have some questions about the comparison results:

1. I can roughly understand why DeiT-tiny shows a large difference between FP and its quantized output — this may come from limited quantization granularity or the nature of the model itself. In that sense, it’s reasonable that the TOSA output would also deviate significantly from FP.
    However, what I’m not fully sure about is **FQ vs TOSA**. In theory, these two should be very close to each other, similar to the other models.
2. For LLaMA, there is already a very large deviation from the FP model **immediately after applying `quantizer.transform_for_annotation`**, even before running FakeQuant or lowering to TOSA.

Is it possible that these models are currently *lowerable*, but numerical correctness is not yet guaranteed for them?

cc @freddan80 @per @zingo @oscarandersson8218 @digantdesai

**Comment by gggekov** [comment]
Hi @hth000 
Thanks for the write-up, interesting problem. 
DeitTiny contains plenty of LayerNorm layers and LayerNorm quantization is quite sensitive to the epsilon value - the default epsilon for LayerNorm is 10ˆ(-5). In the LayerNorm computation, you divide by the sqrt(variance(x)+eps) so with a tiny epsilon, you are amplifying the numerator when the variance is small, e.g. 1/sqrt(10ˆ-5) = 316 and 1/sqrt(10ˆ-1) = 3 = 100x difference. As a result, you amplify the numerator of the LayerNorm and the output tensor gets a really wide range that is hard to quantize in 8x8. We've looked into deit_tiny in the past - you should get improved quantized accuracy if you use a bigger epsilon, e.g. 10^-1. Could you try that? 

Also, could you specify how you calculate the 4 columns in the table, especially the FP vs FQ and FQ vs TOSA columns ? It's curious you get almost identical values.

Also, it's true that the element-wise comparison is interesting, but the most important metric is the accuracy over a dataset - for ex. top1 or top5 accuracy figures for the quantized & FP32 Deit_Tiny graphs. For the quantized case, don't forget to calibrate on real data for meaningful results.

**Comment by hth000** [comment]
Hi @gggekov,

Thanks a lot for your insights and suggestions — they were very helpful. I followed your recommendations and ran a series of experiments, and the results are now much clearer. Below I’ll first explain the **definition of the four columns in the tables**, then discuss the **LayerNorm epsilon experiment**, and finally present results using **real calibration data with accuracy measurements**.

------

## Definition of the four columns

First, here is what each column in the tables represents:

- **FP**
   The original FP32 source model.
- **Decompose with aten**
   The model after ARM quantizer lowering, but **before inserting quantization nodes** (still a floating-point graph).
- **Decompose with quant aten (FQ)**
   Quantization nodes are inserted and calibration is performed to obtain quantization parameters (i.e. fake-quant / quantized graph).
- **TOSA**
   The model is serialized into a `*.tosa` file and inference is run using the `tosa_reference_model` executable.

Regarding the fact that **FP vs FQ** and **FQ vs TOSA** are very close in some cases: I believe this is largely coincidental. Conceptually, I would expect **FP vs FQ** and **FP vs TOSA** to be closer than **FQ vs TOSA**.

------

## LayerNorm epsilon experiment (random calibration)

As you mentioned, LayerNorm quantization is very sensitive to the epsilon value. 

I therefore changed the LayerNorm epsilon from `1e-6` to `0.1`, and evaluated the impact. As expected, increasing epsilon does improve numerical precision when using random calibration data. The results below are obtained using random input for calibration (`torch.rand(1, 3, 224, 224)`).

### Random calibration results

| Model                                                        | **FP vs Decompose with aten**                                | **FP vs Decompose with quant aten (FQ)**                     | **FP vs TOSA**                                               | **Compare FQ vs TOSA**                                       |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **Deit-tiny**  <br />**Layernorm eps: 10e-06-> 0.1**<br /> **Cal:torch.rand(1,3,224,224)** | max  error: 0.0  -> 0.0  <br />mean  error: 0.0  -> 0.0 <br /> cosine  : 1.0 -> 1.0 | max  error: 96.0 -> **55.0** <br /> mean  error: 21.23 ->  14.878  <br />cosine  : 0.8994378800187951          -> **0.9259960372348681** | max  error : 83.0 ->  **66.0**  <br />mean  error: 18.682 ->  16.671  <br />cosine  : 0.9264412351657058         -> 0.9225865455808423 | max  error : 96.0 ->  **62.0**  <br />mean  error: 21.434 ->  **13.843**<br />  cosine  : 0.898735675973962         -> **0.9448885854729189** |

------

## Real calibration data and accuracy evaluation

More importantly, I then ran experiments using **real ImageNet data for calibration**, which I believe is the most meaningful evaluation. I calibrated with **10,240 real images** and evaluated accuracy on a small validation subset (100 images) to quickly sanity-check the behavior.

### Accuracy results (100 images)

- **FP (Decompose with aten)**: 93.0
- **FQ**: 87.0
- **TOSA**: 87.0

These accuracy numbers are consistent with expectations: quantization introduces a reasonable accuracy drop, and TOSA matches the quantized graph closely.

### Precision comparison with real calibration data

| Model                                                        | **FP vs Decompose with aten**                                | **FP vs Decompose with quant aten (FQ)**                     | **FP vs TOSA**                                               | **Compare FQ vs TOSA**                                       |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **Deit-tiny    <br />Layernorm eps: 10e-05<br />Cal: 10240 real images** | max error : 1.0<br />mean error: 0.001<br />cosine  : 0.9999997803740219 | max error : 18.0<br />mean error: 4.178<br />cosine  : 0.9939002122977539 | max error : 20.0<br />mean error: 4.301<br />cosine  : 0.9937475784944306 | max error : 17.0<br />mean error: 3.837<br /><br />cosine  : 0.9946952505634903 |

One somewhat surprising result for me is that **FQ vs TOSA precision becomes quite good once correct calibration data is used**. Initially, I expected that the transformation between FQ and TOSA should not be influenced much by calibration parameters, but empirically it does seem to benefit from proper calibration.

------

## LLaMA support

Finally, regarding LLaMA: I saw the related PRs and it looks like support is still under active development, so it does not appear to be fully supported yet. Please correct me if my understanding is wrong.

PR link for reference:
 https://github.com/pytorch/executorch/pull/15556

Thanks again for the detailed feedback.

**Comment by gggekov** [comment]
Hi @hth000 ,
> One somewhat surprising result for me is that FQ vs TOSA precision becomes quite good once correct calibration data is used. 

I agree that is a bit strange. I wonder if that is not due to a difference in the data used for training vs the random data during calibration. When you calibrate with torch.rand, could you please also add
```
normalize = transforms.Normalize(
    mean=IMAGENET_INCEPTION_MEAN, std=IMAGENET_INCEPTION_STD
)
model_inputs = (normalize(torch.rand((1, 3, 224, 224))),)
```
to make sure the data during PTQ calibration follows the distribution as what the model was trained on ?

**Comment by hth000** [comment]
Hi @gggekov,

Yes, I agree that using data consistent with the original training distribution is more reasonable for PTQ calibration.

I also tried calibrating with random inputs that are normalized using the same ImageNet statistics as the training setup. However, based on the experiments, this change did not lead to a noticeable improvement in precision, regardless of whether the LayerNorm epsilon was kept at the default value or increased to 0.1.

Empirically, it still appears that the FQ vs TOSA precision only becomes consistently close when the quantization parameters are obtained using a sufficiently large amount of real calibration data.
|                                                              | **FP  vs Decompose with** **aten**                           | **FP  vs Decompose with quant** **aten****(FQ)**             | **FP  vs TOSA**                                              | **==  Compare FQ vs TOSA ==**                                |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **Deit-tiny**<br />  **Layernorm** **eps:  10e-05<br />**  Cal:torch.rand(1,3,224,224) | max  error : 1.0  <br />mean  error: 0.002 <br /> cosine  : 0.9999996206158854 | max  error : 66.0  <br />mean  error: 14.046 <br /> cosine  : 0.9410028844788999 | max  error : 86.0  <br />mean  error: 15.125  <br />cosine  : 0.9280324484622701 | max  error : 52.0  <br />mean  error: 11.335  <br />cosine  : 0.9621293365058479 |
| **Deit-tiny** <br /> **Layernorm** **eps:** **10e-05**<br /> Cal:torch.rand(1,3,224,224)<br />  **transformer.Normalize** | max  error : 1.0  <br />mean  error: 0.003  <br />cosine  : 0.9999995414999936 | max  error : 101.0<br />  mean  error: 17.433<br />  cosine  : 0.936096336748049 | max  error : 81.0  <br />mean  error: 15.873  <br />cosine  : 0.9448136131791466 | max  error : 65.0 <br /> mean  error: 10.576 <br /> cosine  : 0.9772789574886362 |
| **Deit-tiny**<br />   **Layernorm** **eps:** **0.1<br />**  Cal: (torch.rand(1,3,224,224))  <br />**transformer.Normalize** | max  error : 0.0  <br />mean  error: 0.0  <br />cosine  : 1.0 | max  error : 88.0  <br />mean  error: 13.093  <br />cosine  : 0.9314040375423385 | max  error : 47.0 <br /> mean  error: 10.678 <br /> cosine  : 0.9584637141963666 | max  error : 57.0  <br />mean  error: 9.133  <br />cosine  : 0.9680621480573112 |
|                                                              |                                                              |                                                              |                                                              |                                                              |
| **Deit-tiny** <br /> **Layernorm** **eps: 10e-05** <br /> **Cal:** **1024** **real images**  <br />**transformer.Normalize** | max  error : 0.0  <br />mean  error: 0.0  <br />cosine  : 1.0000000000000002 | max  error : 61.0  <br />mean  error: 14.611  <br />cosine  : 0.9261875871915061 | max  error : 75.0  <br />mean  error: 14.547<br />  cosine  : 0.9271685728391744 | max  error : 22.0  <br />mean  error: 5.248  <br />cosine  : 0.9908055206889544 |
| **Deit-tiny** <br /> **Layernorm** **eps: 10e-05** <br /> **Cal:** **10240** **real images**  <br />**transformer.Normalize** | max  error : 1.0 <br /> mean  error: 0.001<br />  cosine  : 0.9999997803740219 | max  error : 18.0 <br /> mean  error: 4.178<br />  cosine  : 0.9939002122977539 | max  error : 20.0  <br />mean  error: 4.301  <br />cosine  : 0.9937475784944306 | max  error : 17.0  <br />mean  error: 3.837  <br />cosine  : 0.9946952505634903 |

**Comment by nil-is-all** [comment]
Hi @gggekov is this issue still being looked at?

**Comment by nil-is-all** [comment]
@gggekov gentle ping again
